raiden/network/rpc/client.py

```diff
@@ -914,50 +951,68 @@ def __init__(
             startgas=GAS_LIMIT,
             gasprice=GAS_PRICE,
             poll_timeout=DEFAULT_POLL_TIMEOUT):
-        # pylint: disable=too-many-arguments
 
-        result = jsonrpc_client.call(
-            'eth_getCode',
+        self.address = channel_address
+        self.client = jsonrpc_client
+        self.startgas = startgas
+        self.gasprice = gasprice
+        self.poll_timeout = poll_timeout
+
+        self.client = jsonrpc_client
+        self.node_address = privatekey_to_address(self.client.privkey)
+        self.proxy = jsonrpc_client.new_abi_contract(
+            CONTRACT_MANAGER.get_abi(CONTRACT_NETTING_CHANNEL),
             address_encoder(channel_address),
+        )
+
+        # check we are a participant of the given channel
+        self.detail()
+        self._check_exists()
+
+    def _check_exists(self):
+        result = self.client.call(
+            'eth_getCode',
+            address_encoder(self.address),
             'latest',
         )
 
         if result == '0x':
             raise AddressWithoutCode('Netting channel address {} does not contain code'.format(
-                address_encoder(channel_address),
+                address_encoder(self.address),
             ))
 
-        proxy = jsonrpc_client.new_abi_contract(
-            CONTRACT_MANAGER.get_abi(CONTRACT_NETTING_CHANNEL),
-            address_encoder(channel_address),
-        )
+    def token_address(self):
+        """ Returns the type of token that can be transferred by the channel.
 
-        self.address = channel_address
-        self.proxy = proxy
-        self.client = jsonrpc_client
-        self.startgas = startgas
-        self.gasprice = gasprice
-        self.poll_timeout = poll_timeout
+        Raises:
+            AddressWithoutCode: If the channel was settled prior to the call.
+        """
+        try:
+            address = self.proxy.tokenAddress.call()
+        except:
+            self._check_exists()
```

Comment:
You use this pattern a lot in all of the function so I will use this point to ask about it.

```
try:
    proxy_call()
except:
    check_for_existence()
```

The idea behind this is that you want to catch whatever failure `call()` does, check if the failure was due to empty code and use a specialized exception for it, and if not then propagate the exception. Right?

I suppose that is because `call()` in python would have no way to distinguish between the address having no code or other errors, right?

Reply:
Yes, the idea is to check if the transaction failed if the contract was self destructed.

This is actually unnecessary for `call()`s. I assumed `eth_call` would notify the json rpc caller about a throw and that the python client would raise an error, but I was wrong.